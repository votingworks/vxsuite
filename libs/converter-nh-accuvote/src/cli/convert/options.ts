import { err, ok, Optional, Result } from '@votingworks/basics';
import { safeParse } from '@votingworks/types';
import { promises as fs } from 'fs';
import { isAbsolute, join } from 'path';
import { parseJsonc } from '..';
import {
  ConvertConfig,
  ConvertConfigCard,
  ConvertConfigSchema,
} from '../../convert/types';

/**
 * Options for the `convert` command's default behavior.
 */
export interface ConvertOptions {
  readonly type: 'convert';
  readonly config: ConvertConfig;
}

/**
 * Options for the `help` subcommand.
 */
export interface HelpOptions {
  readonly type: 'help';
}

/**
 * Options for the `convert` command.
 */
export type Options = ConvertOptions | HelpOptions;

/**
 * Parses command-line options for the `convert` command.
 */
export async function parseOptions(
  args: readonly string[]
): Promise<Result<Options, Error>> {
  let configPath: Optional<string>;

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];

    switch (arg) {
      case '-c':
      case '--config': {
        const nextArg = args[i + 1];
        if (nextArg === undefined) {
          return err(new Error('missing path to config file'));
        }
        configPath = nextArg;
        i += 1;
        break;
      }

      case '-h':
      case '--help':
        return ok({ type: 'help' });

      default: {
        return err(new Error(`unknown option: ${arg}`));
      }
    }
  }

  if (!configPath) {
    return err(new Error('missing required config file (--config <file>)'));
  }

  const configText = await fs.readFile(configPath, 'utf8');
  const parseJsonResult = parseJsonc(configText);
  if (parseJsonResult.isErr()) {
    return err(new Error(`error parsing config file: invalid JSON`));
  }

  const configResult = safeParse(ConvertConfigSchema, parseJsonResult.ok());
  if (configResult.isErr()) {
    return err(
      new Error(`error parsing config file: ${configResult.err().message}`)
    );
  }
  const config = configResult.ok();
  const configDir = join(configPath, '..');

  function locatePathRelativeToConfigPath(path: string): string {
    if (isAbsolute(path)) {
      return path;
    }

    return join(configDir, path);
  }

  return ok({
    type: 'convert',
    config: {
      ...config,
      jurisdictions: config.jurisdictions.map((jurisdiction) => ({
        ...jurisdiction,
        cards: jurisdiction.cards.map((card) => ({
          ...card,
          definition: locatePathRelativeToConfigPath(card.definition),
          ballot: locatePathRelativeToConfigPath(card.ballot),
        })) as [ConvertConfigCard, ...ConvertConfigCard[]],
        output: locatePathRelativeToConfigPath(jurisdiction.output),
      })),
    },
  });
}
