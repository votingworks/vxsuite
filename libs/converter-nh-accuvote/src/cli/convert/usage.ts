import { lines } from '@votingworks/basics';
import { unsafeParse } from '@votingworks/types';
import chalk from 'chalk';
import {
  comment,
  number,
  parseJsonc,
  softWrapper,
  string,
  stripAnsi,
} from '..';
import { ConvertConfigSchema } from '../../convert/types';

const EXAMPLE_CONFIG_STRING = `
{
  ${comment('// electionType: general, primary, etc.')}
  ${string('electionType')}: ${string('general')},
  ${string('jurisdictions')}: [
    ${comment('// single-card jurisdiction')}
    {
      ${string('name')}: ${string('Alton')},
      ${string('cards')}: [
        {
          ${string('definition')}: ${string('input/alton/definition.xml')},
          ${string('ballot')}: ${string('input/alton/ballot.pdf')}
        }
      ],
      ${string('output')}: ${string('output/alton')}
    },
    ${comment('// multi-card jurisdiction')}
    {
      ${string('name')}: ${string('Rochester')},
      ${string('cards')}: [
        {
          ${string('definition')}: ${string('input/rochester/card1.xml')},
          ${string('ballot')}: ${string('input/rochester/ballot.pdf')},
          ${comment('// optional: useful for multi-card PDFs')}
          ${string('pages')}: [${number(1)}, ${number(2)}]
        },
        {
          ${string('definition')}: ${string('input/rochester/card2.xml')},
          ${string('ballot')}: ${string('input/rochester/ballot.pdf')},
          ${comment('// optional: useful for multi-card PDFs')}
          ${string('pages')}: [${number(3)}, ${number(4)}]
        }
      ],
      ${string('output')}: ${string('output/rochester')}
    }
  ]
}
`.trim();

/**
 * An example configuration for the convert command.
 *
 * This is primarily here to validate the example config string.
 */
export const EXAMPLE_CONFIG = unsafeParse(
  ConvertConfigSchema,
  parseJsonc(stripAnsi(EXAMPLE_CONFIG_STRING)).unsafeUnwrap()
);

const USAGE = `
Usage: convert --config <config.json>

${chalk.bold('Summary')}

Converts a series of New Hampshire ballot cards into election definitions in the VotingWorks format.

${chalk.bold('Description')}

Uses a config file to define the inputs and outputs (see example below). The config file may include comments (${comment(
  '// like this'
)} or ${comment(
  '/* this */'
)}). Note that if the input data is incorrect then the output will likely be incorrect as well. This tool can correct certain mistakes in the AccuVote XML, such as bubble positions that are slightly off, but cannot correct more egregious errors:

  1. Candidates/contest options in the wrong column.
  2. Candidates/contest options in the wrong order.
  3. Missing timing marks.
  4. Missing contest option bubbles.

The output of this command will include two PDFs for each ballot style, one for printing and one for proofing:
• ${chalk.underline(
  'PRINT PDFs'
)} include a QR code that enables scanning by VxScan or VxCentralScan
• ${chalk.underline('PROOF PDFs')} overlay bubble and contest option locations

${chalk.italic.underline.red(
  `Please review ALL the PDFs labeled PROOF to ensure correctness!`
)}

${chalk.bold('Example Config')}

${EXAMPLE_CONFIG_STRING}
`.trim();

/**
 * Prints the usage information for the `convert` command.
 */
export function usage(out: NodeJS.WritableStream): void {
  for (const line of lines(USAGE).flatMap(softWrapper(80))) {
    out.write(`${line}\n`);
  }
}
