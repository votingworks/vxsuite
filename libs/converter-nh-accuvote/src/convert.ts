/* istanbul ignore file - no logic, just re-exports from other files */
export * from './convert/convert_election_definition';
