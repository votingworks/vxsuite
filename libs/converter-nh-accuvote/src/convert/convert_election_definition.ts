import {
  BallotPageTimingMarkMetadataFront,
  findTemplateGridAndBubbles,
} from '@votingworks/ballot-interpreter';
import {
  assert,
  err,
  ok,
  resultBlock,
  throwIllegalValue,
  typedAs,
} from '@votingworks/basics';
import { Election, GridPosition, getContests } from '@votingworks/types';
import { convertElectionDefinitionHeader } from './convert_election_definition_header';
import { matchContestOptionsOnGrid } from './match_contest_options_on_grid';
import {
  ConvertIssueKind,
  ConvertResult,
  NewHampshireBallotCardDefinition,
  PairColumnEntriesIssueKind,
  TemplateBubbleGridEntry,
  ConvertIssue,
} from './types';

/**
 * Convert New Hampshire XML files to a single {@link Election} object.
 */
export function convertElectionDefinition(
  cardDefinition: NewHampshireBallotCardDefinition
): ConvertResult {
  return resultBlock((fail) => {
    const convertHeader = convertElectionDefinitionHeader(
      cardDefinition.definition
    ).okOrElse(fail);

    const { election, issues: headerIssues } = convertHeader;
    let success = true;
    const issues = [...headerIssues];

    const findTemplateGridAndBubblesResult = findTemplateGridAndBubbles([
      cardDefinition.front,
      cardDefinition.back,
    ]);

    if (findTemplateGridAndBubblesResult.isErr()) {
      return err({
        issues: [
          ...issues,
          typedAs<ConvertIssue>({
            kind: ConvertIssueKind.TimingMarkDetectionFailed,
            message: 'failed to detect timing marks',
            side: 'front',
          }),
        ],
      });
    }

    let [frontGridAndBubbles, backGridAndBubbles] =
      findTemplateGridAndBubblesResult.ok();

    if (
      frontGridAndBubbles.metadata.side === 'back' &&
      backGridAndBubbles.metadata.side === 'front'
    ) {
      [frontGridAndBubbles, backGridAndBubbles] = [
        backGridAndBubbles,
        frontGridAndBubbles,
      ];
    }

    let { paperSize } = election.ballotLayout;

    const frontExpectedPaperSize =
      frontGridAndBubbles.grid.geometry.ballotPaperSize;
    const backExpectedPaperSize =
      backGridAndBubbles.grid.geometry.ballotPaperSize;

    assert(
      frontExpectedPaperSize === backExpectedPaperSize,
      'the paper size should be the same for both sides'
    );

    if (frontExpectedPaperSize !== paperSize) {
      success = frontExpectedPaperSize === backExpectedPaperSize;
      issues.push({
        kind: ConvertIssueKind.InvalidTemplateSize,
        message: `Template images do not match expected sizes. The XML definition says the template images should be "${paperSize}", but the template images are front="${frontExpectedPaperSize}" and back="${backExpectedPaperSize}".`,
        paperSize,
        frontTemplateSize: {
          width: cardDefinition.front.width,
          height: cardDefinition.front.height,
        },
        backTemplateSize: {
          width: cardDefinition.back.width,
          height: cardDefinition.back.height,
        },
      });
      paperSize = frontExpectedPaperSize;
    }

    if (frontGridAndBubbles.metadata.side !== 'front') {
      success = false;
      issues.push({
        kind: ConvertIssueKind.InvalidTimingMarkMetadata,
        message: `front page timing mark metadata is invalid: side=${frontGridAndBubbles.metadata.side}`,
        side: 'front',
        timingMarkBits: frontGridAndBubbles.metadata.bits,
        timingMarks: frontGridAndBubbles.grid.partialTimingMarks,
      });
    }

    if (backGridAndBubbles.metadata.side !== 'back') {
      success = false;
      issues.push({
        kind: ConvertIssueKind.InvalidTimingMarkMetadata,
        message: `back page timing mark metadata is invalid: side=${backGridAndBubbles.metadata.side}`,
        side: 'back',
        timingMarkBits: backGridAndBubbles.metadata.bits,
        timingMarks: backGridAndBubbles.grid.partialTimingMarks,
      });
    }

    if (!success) {
      return err({
        issues,
        election,
      });
    }

    const frontMetadata =
      frontGridAndBubbles.metadata as BallotPageTimingMarkMetadataFront;
    const ballotStyleId = `card-number-${frontMetadata.cardNumber}`;

    const frontTemplateBubbles = frontGridAndBubbles.bubbles;
    const backTemplateBubbles = backGridAndBubbles.bubbles;

    const gridLayout = election.gridLayouts?.[0];
    assert(gridLayout, 'grid layout missing');

    const ballotStyle = election.ballotStyles[0];
    assert(ballotStyle, 'ballot style missing');

    const bubbleGrid = [
      ...frontTemplateBubbles.map<TemplateBubbleGridEntry>((bubble) => ({
        side: 'front',
        column: bubble.x,
        row: bubble.y,
      })),
      ...backTemplateBubbles.map<TemplateBubbleGridEntry>((bubble) => ({
        side: 'back',
        column: bubble.x,
        row: bubble.y,
      })),
    ];

    const pairColumnEntriesResult = matchContestOptionsOnGrid(
      getContests({ ballotStyle, election }),
      gridLayout.gridPositions.map<GridPosition>((gridPosition) => ({
        ...gridPosition,
      })),
      bubbleGrid
    );

    if (pairColumnEntriesResult.isErr()) {
      success = false;

      for (const issue of pairColumnEntriesResult.err().issues) {
        switch (issue.kind) {
          case PairColumnEntriesIssueKind.ColumnCountMismatch:
            issues.push({
              kind: ConvertIssueKind.MismatchedOvalGrids,
              message: `XML definition and ballot images have different number of columns containing ovals: ${issue.columnCounts[0]} vs ${issue.columnCounts[1]}`,
              pairColumnEntriesIssue: issue,
            });
            break;

          case PairColumnEntriesIssueKind.ColumnEntryCountMismatch:
            issues.push({
              kind: ConvertIssueKind.MismatchedOvalGrids,
              message: `XML definition and ballot images have different number of entries in column ${issue.columnIndex}: ${issue.columnEntryCounts[0]} vs ${issue.columnEntryCounts[1]}`,
              pairColumnEntriesIssue: issue,
            });
            break;

          default:
            throwIllegalValue(issue, 'kind');
        }
      }
    }

    const mergedGrids = pairColumnEntriesResult.isOk()
      ? pairColumnEntriesResult.ok().pairs
      : pairColumnEntriesResult.err().pairs;
    const result: Election = {
      ...election,
      ballotLayout: {
        ...election.ballotLayout,
        paperSize,
      },
      ballotStyles: [
        {
          ...ballotStyle,
          id: ballotStyleId,
        },
      ],
      gridLayouts: [
        {
          ...gridLayout,
          columns: frontGridAndBubbles.grid.geometry.gridSize.width,
          rows: frontGridAndBubbles.grid.geometry.gridSize.height,
          ballotStyleId,
          gridPositions: mergedGrids.map(([definition, bubble]) => ({
            ...definition,
            side: bubble.side,
            column: bubble.column,
            row: bubble.row,
          })),
        },
      ],
    };

    return ok({ issues, election: result });
  });
}
