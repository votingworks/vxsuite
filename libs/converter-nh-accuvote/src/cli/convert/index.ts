import {
  loadImage,
  toImageData,
  imageDebugger,
} from '@votingworks/image-utils';
import { err, ok, Result } from '@votingworks/basics';
import { DOMParser } from '@xmldom/xmldom';
import { enable as enableDebug } from 'debug';
import { promises as fs } from 'fs';
import { basename } from 'path';
import { RealIo, Stdio } from '..';
import { convertElectionDefinition } from '../../convert/convert_election_definition';
import * as templates from '../../data/templates';
import { NewHampshireBallotCardDefinition } from '../../convert/types';

interface ConvertOptions {
  readonly type: 'convert';
  readonly definitionPath: string;
  readonly frontBallotPath: string;
  readonly backBallotPath: string;
  readonly outputPath?: string;
  readonly debug: boolean;
}

interface HelpOptions {
  readonly type: 'help';
}

type Options = ConvertOptions | HelpOptions;

/**
 * Parses {@link xml} and returns the root element.
 */
function parseXml(xml: string): Element {
  return new DOMParser().parseFromString(xml, 'application/xml')
    .documentElement;
}

function parseOptions(args: readonly string[]): Result<Options, Error> {
  let definitionPath: string | undefined;
  let frontBallotPath: string | undefined;
  let backBallotPath: string | undefined;
  let outputPath: string | undefined;
  let debug = false;

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];

    switch (arg) {
      case '-o':
      case '--output': {
        const nextArg = args[i + 1];
        if (nextArg === undefined) {
          return err(new Error(`missing output path after ${arg}`));
        }
        // '-' is a special case for stdout, which is the default
        if (nextArg !== '-') {
          outputPath = nextArg;
        }
        i += 1;
        break;
      }

      case '--debug': {
        debug = true;
        break;
      }

      case '-h':
      case '--help':
        return ok({ type: 'help' });

      default: {
        if (arg?.startsWith('-')) {
          return err(new Error(`unknown option: ${arg}`));
        }

        if (arg?.endsWith('.xml')) {
          definitionPath = arg;
        } else if (arg?.endsWith('.jpeg') || arg?.endsWith('.jpg')) {
          if (frontBallotPath) {
            backBallotPath = arg;
          } else {
            frontBallotPath = arg;
          }
        } else {
          return err(new Error(`unexpected argument: ${arg}`));
        }
      }
    }
  }

  if (!definitionPath) {
    return err(new Error('missing definition path'));
  }

  if (!frontBallotPath) {
    return err(new Error('missing front ballot path'));
  }

  if (!backBallotPath) {
    return err(new Error('missing back ballot path'));
  }

  return ok({
    type: 'convert',
    definitionPath,
    frontBallotPath,
    backBallotPath,
    outputPath,
    debug,
  });
}

function usage(out: NodeJS.WritableStream): void {
  out.write(
    `usage: convert <definition.xml> <front-ballot.jpg> <back-ballot.jpg> [-o <output.json>] [--debug]\n`
  );
}

/**
 * Converts New Hampshire (NH) ballot data to the VotingWorks format.
 */
export async function main(
  args: readonly string[],
  io: Stdio = RealIo
): Promise<number> {
  const parseResult = parseOptions(args);

  if (parseResult.isErr()) {
    io.stderr.write(`error: ${parseResult.err().message}\n`);
    return 1;
  }

  const options = parseResult.ok();

  if (options.type === 'help') {
    usage(io.stdout);
    return 0;
  }

  if (options.debug) {
    enableDebug('converter-nh-accuvote:*');
  }

  const { definitionPath, frontBallotPath, backBallotPath, outputPath } =
    options;

  const definitionContent = await fs.readFile(definitionPath, 'utf8');
  const frontBallotImage = toImageData(await loadImage(frontBallotPath));
  const backBallotImage = toImageData(await loadImage(backBallotPath));

  const cardDefinition: NewHampshireBallotCardDefinition = {
    definition: parseXml(definitionContent),
    front: frontBallotImage,
    back: backBallotImage,
  };

  const convertResult = convertElectionDefinition(cardDefinition, {
    ovalTemplate: await templates.getOvalTemplate(),
    debug: imageDebugger(
      outputPath ??
        `convert-front=${basename(frontBallotPath)}-back=${basename(
          backBallotPath
        )}`,
      { width: frontBallotImage.width, height: frontBallotImage.height }
    ),
  });

  const { issues = [] } = convertResult.isOk()
    ? convertResult.ok()
    : convertResult.err();

  if (issues.length > 0) {
    io.stderr.write(convertResult.isOk() ? 'warning: ' : 'error: ');
    io.stderr.write(`conversion completed with issues:\n`);
    for (const issue of issues) {
      io.stderr.write(`- ${issue.message}\n`);
    }
  }

  if (convertResult.isOk()) {
    const { election } = convertResult.ok();
    if (election) {
      const output = JSON.stringify(election, null, 2);
      if (!outputPath) {
        io.stdout.write(output);
      } else {
        await fs.writeFile(outputPath, output);
      }
    }
  }

  return convertResult.isOk() ? 0 : 1;
}
