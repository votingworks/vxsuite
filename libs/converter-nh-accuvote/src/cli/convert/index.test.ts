import { electionGridLayoutNewHampshireHudsonFixtures } from '@votingworks/fixtures';
import { mockReadable, mockWritable, mockOf } from '@votingworks/test-utils';
import { createImageData } from 'canvas';
import { readFileSync } from 'fs';
import { fileSync } from 'tmp';
import { err, ok } from '@votingworks/basics';
import { main } from '.';
import { Stdio } from '..';
import { convertElectionDefinition } from '../../convert/convert_election_definition';
import { ConvertIssueKind } from '../../convert/types';

jest.mock('../../convert/convert_election_definition');
jest.mock(
  '@votingworks/image-utils',
  (): Partial<typeof import('@votingworks/image-utils')> => ({
    loadImage: jest.fn(),
    loadImageData: jest.fn().mockReturnValue(createImageData(1, 1)),
    toImageData: jest.fn().mockReturnValue(createImageData(1, 1)),
  })
);

test('--help', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  expect(await main(['--help'], io)).toEqual(0);

  expect(io.stdout.toString()).toMatchInlineSnapshot(`
    "usage: convert <definition.xml> <front-ballot.jpg> <back-ballot.jpg> [-o <output.json>] [--debug]
    "
  `);
});

test('-h', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  expect(await main(['-h'], io)).toEqual(0);

  expect(io.stdout.toString()).toMatchInlineSnapshot(`
    "usage: convert <definition.xml> <front-ballot.jpg> <back-ballot.jpg> [-o <output.json>] [--debug]
    "
  `);
});

test('missing output after --output', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(
    [
      electionGridLayoutNewHampshireHudsonFixtures.definitionXml.asFilePath(),
      'front.jpeg',
      'back.jpeg',
      '--output',
    ],
    io
  );

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: missing output path after --output
    "
  `);
  expect(exitCode).toEqual(1);
});

test('unexpected option', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(['--nope'], io);

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: unknown option: --nope
    "
  `);
  expect(exitCode).toEqual(1);
});

test('unexpected argument', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(
    ['definition.xml', 'front.jpeg', 'back.jpeg', 'what-is-this.json'],
    io
  );

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: unexpected argument: what-is-this.json
    "
  `);
  expect(exitCode).toEqual(1);
});

test('missing definition path', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(['front.jpeg', 'back.jpeg'], io);

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: missing definition path
    "
  `);
  expect(exitCode).toEqual(1);
});

test('missing front ballot path', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(['definition.xml'], io);

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: missing front ballot path
    "
  `);
  expect(exitCode).toEqual(1);
});

test('missing back ballot path', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const exitCode = await main(['definition.xml', 'front.jpeg'], io);

  expect(io.stderr.toString()).toMatchInlineSnapshot(`
    "error: missing back ballot path
    "
  `);
  expect(exitCode).toEqual(1);
});

test('convert to stdout', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const { election } = electionGridLayoutNewHampshireHudsonFixtures;

  mockOf(convertElectionDefinition).mockReturnValue(
    ok({ election, issues: [] })
  );

  const exitCode = await main(
    [
      electionGridLayoutNewHampshireHudsonFixtures.definitionXml.asFilePath(),
      'front.jpeg',
      'back.jpeg',
      '-o',
      '-',
    ],
    io
  );

  expect({
    exitCode,
    stdout: io.stdout.toString(),
    stderr: io.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: JSON.stringify(election, null, 2),
    stderr: '',
  });
});

test('convert to file', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  const { election } = electionGridLayoutNewHampshireHudsonFixtures;

  mockOf(convertElectionDefinition).mockReturnValue(
    ok({ election, issues: [] })
  );

  const outputFile = fileSync();
  const exitCode = await main(
    [
      electionGridLayoutNewHampshireHudsonFixtures.definitionXml.asFilePath(),
      'front.jpeg',
      'back.jpeg',
      '-o',
      outputFile.name,
    ],
    io
  );

  expect({
    exitCode,
    stdout: io.stdout.toString(),
    stderr: io.stderr.toString(),
  }).toEqual({
    exitCode: 0,
    stdout: '',
    stderr: '',
  });

  expect(readFileSync(outputFile.name, 'utf-8')).toEqual(
    JSON.stringify(election, null, 2)
  );
});

test('convert fails', async () => {
  const io: Stdio = {
    stdin: mockReadable(),
    stdout: mockWritable(),
    stderr: mockWritable(),
  };

  mockOf(convertElectionDefinition).mockReturnValue(
    err({
      issues: [
        {
          kind: ConvertIssueKind.MissingDefinitionProperty,
          message: 'ElectionID is missing',
          property: 'AVSInterface > AccuvoteHeaderInfo > ElectionID',
        },
      ],
    })
  );

  const exitCode = await main(
    [
      electionGridLayoutNewHampshireHudsonFixtures.definitionXml.asFilePath(),
      'front.jpeg',
      'back.jpeg',
      '-o',
      '-',
    ],
    io
  );

  expect({
    exitCode,
    stdout: io.stdout.toString(),
    stderr: io.stderr.toString(),
  }).toEqual({
    exitCode: 1,
    stdout: '',
    stderr:
      'error: conversion completed with issues:\n- ElectionID is missing\n',
  });
});
