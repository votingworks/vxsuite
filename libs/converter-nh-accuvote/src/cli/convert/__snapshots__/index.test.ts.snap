// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`help: --help 1`] = `
"Usage: convert --config <config.json>

Summary

Converts a series of New Hampshire ballot cards into election definitions in the
VotingWorks format.

Description

Uses a config file to define the inputs and outputs (see example below). The
config file may include comments (// like this or /* this */). Note that if the
input data is incorrect then the output will likely be incorrect as well. This
tool can correct certain mistakes in the AccuVote XML, such as bubble positions
that are slightly off, but cannot correct more egregious errors:

  1. Candidates/contest options in the wrong column.
  2. Candidates/contest options in the wrong order.
  3. Missing timing marks.
  4. Missing contest option bubbles.

The output of this command will include two PDFs for each ballot style, one for
printing and one for proofing:
• PRINT PDFs include a QR code that enables scanning by VxScan or VxCentralScan
• PROOF PDFs overlay bubble and contest option locations

Please review ALL the PDFs labeled PROOF to ensure correctness!

Example Config

{
  // electionType: general, primary, etc.
  "electionType": "general",
  "jurisdictions": [
    // single-card jurisdiction
    {
      "name": "Alton",
      "cards": [
        {
          "definition": "input/alton/definition.xml",
          "ballot": "input/alton/ballot.pdf"
        }
      ],
      "output": "output/alton"
    },
    // multi-card jurisdiction
    {
      "name": "Rochester",
      "cards": [
        {
          "definition": "input/rochester/card1.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [1, 2]
        },
        {
          "definition": "input/rochester/card2.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [3, 4]
        }
      ],
      "output": "output/rochester"
    }
  ]
}
"
`;

exports[`help: -h 1`] = `
"Usage: convert --config <config.json>

Summary

Converts a series of New Hampshire ballot cards into election definitions in the
VotingWorks format.

Description

Uses a config file to define the inputs and outputs (see example below). The
config file may include comments (// like this or /* this */). Note that if the
input data is incorrect then the output will likely be incorrect as well. This
tool can correct certain mistakes in the AccuVote XML, such as bubble positions
that are slightly off, but cannot correct more egregious errors:

  1. Candidates/contest options in the wrong column.
  2. Candidates/contest options in the wrong order.
  3. Missing timing marks.
  4. Missing contest option bubbles.

The output of this command will include two PDFs for each ballot style, one for
printing and one for proofing:
• PRINT PDFs include a QR code that enables scanning by VxScan or VxCentralScan
• PROOF PDFs overlay bubble and contest option locations

Please review ALL the PDFs labeled PROOF to ensure correctness!

Example Config

{
  // electionType: general, primary, etc.
  "electionType": "general",
  "jurisdictions": [
    // single-card jurisdiction
    {
      "name": "Alton",
      "cards": [
        {
          "definition": "input/alton/definition.xml",
          "ballot": "input/alton/ballot.pdf"
        }
      ],
      "output": "output/alton"
    },
    // multi-card jurisdiction
    {
      "name": "Rochester",
      "cards": [
        {
          "definition": "input/rochester/card1.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [1, 2]
        },
        {
          "definition": "input/rochester/card2.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [3, 4]
        }
      ],
      "output": "output/rochester"
    }
  ]
}
"
`;

exports[`missing config file after --config 1`] = `
"Error: missing path to config file
Usage: convert --config <config.json>

Summary

Converts a series of New Hampshire ballot cards into election definitions in the
VotingWorks format.

Description

Uses a config file to define the inputs and outputs (see example below). The
config file may include comments (// like this or /* this */). Note that if the
input data is incorrect then the output will likely be incorrect as well. This
tool can correct certain mistakes in the AccuVote XML, such as bubble positions
that are slightly off, but cannot correct more egregious errors:

  1. Candidates/contest options in the wrong column.
  2. Candidates/contest options in the wrong order.
  3. Missing timing marks.
  4. Missing contest option bubbles.

The output of this command will include two PDFs for each ballot style, one for
printing and one for proofing:
• PRINT PDFs include a QR code that enables scanning by VxScan or VxCentralScan
• PROOF PDFs overlay bubble and contest option locations

Please review ALL the PDFs labeled PROOF to ensure correctness!

Example Config

{
  // electionType: general, primary, etc.
  "electionType": "general",
  "jurisdictions": [
    // single-card jurisdiction
    {
      "name": "Alton",
      "cards": [
        {
          "definition": "input/alton/definition.xml",
          "ballot": "input/alton/ballot.pdf"
        }
      ],
      "output": "output/alton"
    },
    // multi-card jurisdiction
    {
      "name": "Rochester",
      "cards": [
        {
          "definition": "input/rochester/card1.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [1, 2]
        },
        {
          "definition": "input/rochester/card2.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [3, 4]
        }
      ],
      "output": "output/rochester"
    }
  ]
}
"
`;

exports[`unexpected option 1`] = `
"Error: unknown option: --nope
Usage: convert --config <config.json>

Summary

Converts a series of New Hampshire ballot cards into election definitions in the
VotingWorks format.

Description

Uses a config file to define the inputs and outputs (see example below). The
config file may include comments (// like this or /* this */). Note that if the
input data is incorrect then the output will likely be incorrect as well. This
tool can correct certain mistakes in the AccuVote XML, such as bubble positions
that are slightly off, but cannot correct more egregious errors:

  1. Candidates/contest options in the wrong column.
  2. Candidates/contest options in the wrong order.
  3. Missing timing marks.
  4. Missing contest option bubbles.

The output of this command will include two PDFs for each ballot style, one for
printing and one for proofing:
• PRINT PDFs include a QR code that enables scanning by VxScan or VxCentralScan
• PROOF PDFs overlay bubble and contest option locations

Please review ALL the PDFs labeled PROOF to ensure correctness!

Example Config

{
  // electionType: general, primary, etc.
  "electionType": "general",
  "jurisdictions": [
    // single-card jurisdiction
    {
      "name": "Alton",
      "cards": [
        {
          "definition": "input/alton/definition.xml",
          "ballot": "input/alton/ballot.pdf"
        }
      ],
      "output": "output/alton"
    },
    // multi-card jurisdiction
    {
      "name": "Rochester",
      "cards": [
        {
          "definition": "input/rochester/card1.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [1, 2]
        },
        {
          "definition": "input/rochester/card2.xml",
          "ballot": "input/rochester/ballot.pdf",
          // optional: useful for multi-card PDFs
          "pages": [3, 4]
        }
      ],
      "output": "output/rochester"
    }
  ]
}
"
`;
